proposals/caller-argument-expression.md

```diff
@@ -0,0 +1,245 @@
+  - In the case the attribute's effects are not enabled, calling methods marked with the attribute would not be an error, to allow existing methods to use the attribute and maintain source compatibility. However, the attribute would be ignored and the method would be called with whatever default value was provided.
+
+```cs
+// Assembly1
+
+void Foo(string bar); // V1
+void Foo(string bar, string barExpression = "not provided"); // V2
+void Foo(string bar, [CallerArgumentExpression("bar")] string barExpression = "not provided"); // V3
+
+// Assembly2
+
+Foo(a); // V1: Compiles to Foo(a), V2, V3: Compiles to Foo(a, "not provided")
+Foo(a, "provided"); // V2, V3: Compiles to Foo(a, "provided")
+
+// Assembly3
+
+[assembly: EnableCallerArgumentExpression]
+
+Foo(a); // V1: Compiles to Foo(a), V2: Compiles to Foo(a, "not provided"), V3: Compiles to Foo(a, "a")
+Foo(a, "provided"); // V2, V3: Compiles to Foo(a, "provided")
+```
+
+- To prevent the [binary compatibility problem][drawbacks] from occurring every time we want to add new caller info to `Debug.Assert`, an alternative solution would be to add a `CallerInfo` struct to the framework that contains all the necessary information about the caller.
+
+```cs
+struct CallerInfo
+{
+    public string MemberName { get; set; }
+    public string TypeName { get; set; }
+    public string Namespace { get; set; }
+    public string FullTypeName { get; set; }
+    public string FilePath { get; set; }
+    public int LineNumber { get; set; }
+    public int ColumnNumber { get; set; }
+    public Type Type { get; set; }
+    public MethodBase Method { get; set; }
+    public string[] ArgumentExpressions { get; set; }
+}
+
+[Flags]
+enum CallerInfoOptions
+{
+    MemberName = 1, TypeName = 2, ...
+}
+
+public static class Debug
+{
+    public static void Assert(bool condition,
+        // If a flag is not set here, the corresponding CallerInfo member is not populated by the caller, so it's
+        // pay-for-play friendly.
+        [CallerInfo(CallerInfoOptions.FilePath | CallerInfoOptions.Method | CallerInfoOptions.ArgumentExpressions)] CallerInfo callerInfo = default(CallerInfo))
+    {
+        string filePath = callerInfo.FilePath;
+        MethodBase method = callerInfo.Method;
+        string conditionExpression = callerInfo.ArgumentExpressions[0];
+        ...
+    }
+}
+
+class Bar
+{
+    void Foo()
+    {
+        Debug.Assert(false);
+
+        // Translates to:
+
+        var callerInfo = new CallerInfo();
+        callerInfo.FilePath = @"C:\Bar.cs";
+        callerInfo.Method = MethodBase.GetCurrentMethod();
+        callerInfo.ArgumentExpressions = new string[] { "false" };
+        Debug.Assert(false, callerInfo);
+    }
+}
+```
+
+This was originally proposed at https://github.com/dotnet/csharplang/issues/87.
```

Comment:
@HaloFour, I liked your idea of having a `CallerInfo` struct to enscapulate all the necessary information about the caller. I took your idea and expanded upon it a bit in the alternatives section of this proposal.